segmentation.py

```diff
@@ -0,0 +1,575 @@
+
+            # Accumulate the log-likelihood of the data.
+            # (Equation 17)
+            hazard = self.__ratefun(hypotheses['count'])
+            aux = np.log(hazard) + log_density + hypotheses['log_probability']
+            loglik = self.__accum(loglik, aux)
+
+            # Keep track of the highest, log-likelihood.
+            if aux > logmax:
+                logmax, ind = aux, hypotheses['count']
+
+            # Update and accumulate the log-probabilities.
+            hypotheses['log_probability'] += np.log1p(-hazard) + log_density
+            logsum = self.__accum(logsum, hypotheses['log_probability'])
+
+        stat = MatrixVariateNormalInvGamma(self.__mu,
+                                           self.__omega,
+                                           self.__sigma,
+                                           self.__eta)
+
+        # Add a new hypothesis, which states that the next segment is about to
+        # begin.
+        self.__hypotheses.append({'index': self.__update_count,
+                                  'count': 0,
+                                  'log_probability': loglik,
+                                  'distribution': stat,
+                                  'log_constant': stat.logconst()})
+
+        logsum = self.__accum(logsum, loglik)
+        self.logsum = logsum
+
+        # Normalize the hypotheses so that their probabilities sum to one.
+        for hypothesis in self.__hypotheses:
+            hypothesis['log_probability'] -= logsum
+
+        # Automatically trim hypothesis on each update if requested.
+        if self.__maximum_hypotheses is not None:
+            self.trim(minprob=self.__minimum_probability,
+                      maxhypot=self.__maximum_hypotheses)
+
+        # In the max-product algorithm, keep track of the most likely
+        # hypotheses.
+        if self.__alg__ == 'maxprod':
+            loglik = logmax
+            self.__index.append(ind)
+
+        # In the sum-product algorithm, keep track of the probabilities.
+        else:
+            iteration = list()
+            for hypothesis in self.__hypotheses:
+                iteration.append((hypothesis['count'],
+                                  hypothesis['log_probability']))
+
+            self.__probabilities.append(iteration)
+
+    def trim(self, minprob=1.0e-6, maxhypot=20):
+        """Prune hypotheses to limit computational complexity.
+
+        The computational complexity of the algorithm can be managed by
+        limiting the number of hypotheses maintained. This method limits the
+        number of hypotheses maintained by the model by:
+
+            1) Removing any hypotheses with a support (probability) less than
+               ``minprob``.
+
+            2) Preserving the first ``maxhypot`` likely hypotheses and
+               discarding the rest.
+
+        """
+
+        # Skip pruning if less hypotheses exist than the maximum allowed.
+        if len(self.__hypotheses) <= maxhypot:
```

Comment:
I guess we should also review variable naming at some point (`max_hypotheses` instead of `maxhypot`). But this'll be in a future pull request.